File: app/src/main/java/com/pomo/sync/identity/AdmissionSession.kt
package com.pomo.sync.identity

import com.pomo.sync.protocol.ProtocolBytes

internal data class AdmissionSnapshot(
    val memberId: ProtocolBytes,
    val admissionId: ProtocolBytes,
    val deviceId: ProtocolBytes,
    val transcriptHash: ProtocolBytes,
    val stage: AdmissionStage,
)

/** Resumable admission state. Callers persist [snapshot] after every successful transition. */
internal class AdmissionSession private constructor(
    private var current: AdmissionSnapshot,
) {
    fun snapshot(): AdmissionSnapshot = current.copy()

    fun verifyFingerprints(
        memberId: ProtocolBytes,
        deviceId: ProtocolBytes,
        transcriptHash: ProtocolBytes,
    ) {
        check(current.stage == AdmissionStage.OFFER_CREATED)
        require(memberId == current.memberId) { "Member fingerprint mismatch" }
        require(deviceId == current.deviceId) { "Device fingerprint mismatch" }
        require(transcriptHash == current.transcriptHash) { "Admission transcript mismatch" }
        current = current.copy(stage = AdmissionStage.MUTUAL_FINGERPRINT_VERIFIED)
    }

    fun advance(next: AdmissionStage) {
        require(current.stage != AdmissionStage.IDENTITY_BLOCKED) { "Blocked identity cannot continue admission" }
        val linear = AdmissionStage.entries.filter { it != AdmissionStage.IDENTITY_BLOCKED }
        val expected = linear.getOrNull(linear.indexOf(current.stage) + 1)
        require(next == expected) { "Admission stages cannot be skipped or rewound" }
        current = current.copy(stage = next)
    }

    fun blockDifferentMember() {
        require(current.stage !in postAuthorizationStages) {
            "After authorization, cancel becomes revocation"
        }
        current = current.copy(stage = AdmissionStage.IDENTITY_BLOCKED)
    }

    companion object {
        fun create(snapshot: AdmissionSnapshot): AdmissionSession {
            require(snapshot.stage == AdmissionStage.OFFER_CREATED)
            return AdmissionSession(snapshot)
        }

        fun resume(snapshot: AdmissionSnapshot): AdmissionSession = AdmissionSession(snapshot)

        private val postAuthorizationStages =
            setOf(
                AdmissionStage.AUTHORIZATION_COMMITTED,
                AdmissionStage.BASELINE_VERIFIED,
                AdmissionStage.READY_ACK_COMMITTED,
            )
    }
}

File: extension/src/sync/identity/AdmissionSession.ts
import type { AdmissionStage } from "./types";

export interface AdmissionSnapshot {
  readonly memberId: string;
  readonly admissionId: string;
  readonly deviceId: string;
  readonly transcriptHash: string;
  readonly stage: AdmissionStage;
}

const STAGES: readonly AdmissionStage[] = [
  "OFFER_CREATED",
  "MUTUAL_FINGERPRINT_VERIFIED",
  "AUTHORIZATION_COMMITTED",
  "BASELINE_VERIFIED",
  "READY_ACK_COMMITTED",
];

/** Persist the returned snapshot in syncAdmissions after every transition. */
export class AdmissionSession {
  #current: AdmissionSnapshot;

  constructor(snapshot: AdmissionSnapshot) {
    this.#current = { ...snapshot };
  }

  snapshot(): AdmissionSnapshot {
    return { ...this.#current };
  }

  verifyFingerprints(memberId: string, deviceId: string, transcriptHash: string): void {
    if (this.#current.stage !== "OFFER_CREATED") throw new Error("Admission fingerprint step already completed");
    if (memberId !== this.#current.memberId) throw new Error("Member fingerprint mismatch");
    if (deviceId !== this.#current.deviceId) throw new Error("Device fingerprint mismatch");
    if (transcriptHash !== this.#current.transcriptHash) throw new Error("Admission transcript mismatch");
    this.#current = { ...this.#current, stage: "MUTUAL_FINGERPRINT_VERIFIED" };
  }

  advance(next: AdmissionStage): void {
    const index = STAGES.indexOf(this.#current.stage);
    if (STAGES[index + 1] !== next) throw new Error("Admission stages cannot be skipped or rewound");
    this.#current = { ...this.#current, stage: next };
  }
}
